### ISSUE TITLE
[TPU] Bug: Reverse is orders of magnitude slower on TPU

### ISSUE THREAD

**Issue by bjenik** [openxla/xla#23191]
The following sample gets around 1400it/s on H100 and 11it/s on v6e. It seems there's an issue with the TPU implementation of [reverse](https://openxla.org/xla/operation_semantics#rev_reverse). This example is already isolated, in practice it gets generated when calling irffft.

```
import os 
os.environ["XLA_FLAGS"] = "--xla_dump_to=./hlo"
from tqdm import tqdm
import jax 
import jax.numpy as jnp
from jax.experimental import mesh_utils
from jax.sharding import Mesh, NamedSharding, PartitionSpec

num_devices = jax.device_count()
mesh = Mesh(mesh_utils.create_device_mesh((jax.device_count(),)), ("batch",))

@jax.jit
def rp1(data):
    data = jax.lax.with_sharding_constraint(data, NamedSharding(mesh, PartitionSpec("batch")))
    data = data[:,:,:,:,:,::-1] + 1
    data = jax.lax.with_sharding_constraint(data, NamedSharding(mesh, PartitionSpec("batch")))
    return data

@jax.jit
def make_data():
    data = jnp.ones((num_devices * 32, 64, 16, 16, 8, 31))
    data = jax.lax.with_sharding_constraint(data, NamedSharding(mesh, PartitionSpec("batch")))
    return data

data = make_data()
with jax.profiler.trace("./tensorboard"):
    for i in tqdm(range(1000)):
        data = rp1(data)

```

**Comment by rdyro** [comment]
I experimented with two alternatives:

* `data = jnp.flip(data, axis=-1) + 1` - ~ 50 ms

![Image](https://github.com/user-attachments/assets/f0caaaf7-66e5-4826-914d-a433076212b1)

* and a gather: `data = data[..., jnp.arange(data.shape[-1])[::-1]] + 1` ~ 6 ms

![Image](https://github.com/user-attachments/assets/fdb19e66-a2cd-4c7b-8f69-50b6dc27c55a)

The gather generates a while loop which seems to be much more efficient. 

One speculation is maybe the reverse operation is VMEM optimized, but your array size is large (500 MB) so needs to use HBM directly and ends up being inefficient - but I'm curious why the compiler does replace the gather, but not the reverse with a more optimized version!

**Comment by rdyro** [comment]
An important detail: I tested it on a v5e so far, and there the HBM BW suggests 2.5 ms, so I think the optimized gather is quite close to it (although it'd ideally be fused with the scalar add in this repro.

Yes, that lowering is problematic... What's the performance like when you implement irfft yourself with the gather trick and ifft? I'll work on creating an internal issue.

**Comment by bjenik** [comment]
Re-implementing irfft to avoid the reverse works as a bandaid in the meantime - thanks. It's probably still better to have this fixed as others may hit it as well. I'm also personally curious what is actually happening on the TPU to cause this level of overhead.

**Comment by rdyro** [comment]
> Re-implementing irfft to avoid the reverse works as a bandaid in the meantime - thanks. 

That's good that an immediate workaround is possible, if you don't mind, do you have a short code snippet for this just as a public reference?

>It's probably still better to have this fixed as others may hit it as well. I'm also personally curious what is actually happening on the TPU to cause this level of overhead.

Without a doubt! I'm following up on this internally, including if I can get an explanation for this.

**Comment by bjenik** [comment]
Seeing 500it/s now on v6e, which is approximately in line with expectations given HBM bandwidth, so this appears to be fixed - thanks!

PS: somehow I didn't get any notification that there was progress so - only saw it randomly checking now - sorry about that.

**Comment by rdyro** [comment]
Thanks for checking, it's good it's as expected now!

**Comment by penumallas1261** [comment]
Hi @bjenik ,

I am closing this issue as it is fixed in latest code of XLA. Please feel free to reopen the issue if you still have a concern. Thanks!
